nvtabular/loader/tensorflow.py

```diff
@@ -416,8 +422,48 @@ def _get_sparse_tensor(self, values, indices, num_rows, seq_limit):
         return sparse_tensor
 
     def _build_sparse_tensor(self, values, offsets, diff_offsets, num_rows, seq_limit):
+        """Builds sparse tensors in the TensorFlow dataloader.
+
+        Parameters
+        ----------
+        values :
+            The values to build our intermediate ragged tensor from.
+        offsets :
+            Not currently used.
+        diff_offsets :
+            The row lengths to build our intermediate ragged tensor from.
+        num_rows :
+            Not currently used.
+        seq_limit : int
+            The desired length of output sequences.
+
+        Returns
+        -------
+        tf.sparse
+            The built TensorFlow sparse tensor.
+
+        """
         ragged = tf.RaggedTensor.from_row_lengths(values=values, row_lengths=diff_offsets)
-        tensor = tf.RaggedTensor.from_tensor(ragged.to_tensor(shape=[None, seq_limit])).to_sparse()
+
+        # Get vector of padding lengths using tf ops like reduce_sum.
+        non_zero_entries_by_row = tf.math.reduce_sum(ragged / ragged, axis=1)
+        paddings = seq_limit - non_zero_entries_by_row.numpy()
+
+        # Make zeros ragged tensor to pad our data tensor with.
+        total_entries = ragged.shape[0] * seq_limit
+        non_zero_entries = tf.reduce_sum(ragged / ragged).numpy()
+        zeros_count = total_entries - non_zero_entries
+        zeros_values = tf.zeros(shape=(int(zeros_count)), dtype=tf.dtypes.int64)
+        zeros = tf.RaggedTensor.from_row_lengths(values=zeros_values, row_lengths=paddings)
+
+        # Concatenate zeros ragged tensor with our data tensor on either the left or the right,
+        # depending on either left_pad or not.
+        if self.pad_left:
+            tensor = tf.concat([zeros, ragged], axis=1).to_tensor()
+        else:
+            tensor = tf.concat([ragged, zeros], axis=1).to_tensor()
+
+        tensor = tf.RaggedTensor.from_tensor(tensor).to_sparse()
```

Comment:
@jperez999 Are you able to see this?

Reply:
I see it now... Can you time this compared to the example I gave... would like to see the timing on this execution.

Reply:
This is good to hear. Yes one moment please. Is there a particular timing framework or method that you use, for consistency of the timings?

Reply:
For this current code block using `tf.concat()`, `pytest --durations=0 -vv tests/unit/loader/test_tf_dataloader.py -k "test_sparse_tensor_"` gives:

<details>
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /nvtabular, configfile: pyproject.toml
collected 77 items / 75 deselected / 2 selected                                                                                                                                                                   

tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[False] PASSED                                                                                                                      [ 50%]
tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[True] PASSED                                                                                                                       [100%]

================================================================================================ slowest durations ================================================================================================
1.38s call     tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[False]
0.09s call     tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[True]
0.00s setup    tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[True]
0.00s setup    tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[False]
0.00s teardown tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[False]
0.00s teardown tests/unit/loader/test_tf_dataloader.py::test_sparse_tensor_left_padding[True]
======================================================================================== 2 passed, 75 deselected in 3.18s =========================================================================================
</details>

Reply:
Do you have a commit with the `tf.reverse()` code that I can checkout and do this same timing on?

Reply:
yeah I just meant grabbing that same mini tensor I create in my example (called digits) and run it through your scenario the idea is to get a comparison to see which is the fastest method to ensure that is what we select. 

```
        ragged = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
        ### from here down your code ###
        non_zero_entries_by_row = tf.math.reduce_sum(ragged / ragged, axis=1)
        paddings = seq_limit - non_zero_entries_by_row.numpy()

        # Make zeros ragged tensor to pad our data tensor with.
        total_entries = ragged.shape[0] * seq_limit
        non_zero_entries = tf.reduce_sum(ragged / ragged).numpy()
        zeros_count = total_entries - non_zero_entries
        zeros_values = tf.zeros(shape=(int(zeros_count)), dtype=tf.dtypes.int64)
        zeros = tf.RaggedTensor.from_row_lengths(values=zeros_values, row_lengths=paddings)

        # Concatenate zeros ragged tensor with our data tensor on either the left or the right,
        # depending on either left_pad or not.
        if self.pad_left:
            tensor = tf.concat([zeros, ragged], axis=1).to_tensor()
        else:
            tensor = tf.concat([ragged, zeros], axis=1).to_tensor()
```

Reply:
This is the code I reported the ~.002 seconds execution time
```
igits = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
max_len = max(max(len(row) for row in digits), 7)
tensor = tf.stack([tf.pad(row, [[max_len - len(row), 0]]) for row in digits], axis=0);
```

Reply:
I see, here are the timings I got. 

```
import tensorflow as tf

"""
: %timeit foo()
1.24 ms ± 15.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

: %timeit bar()
24.6 ms ± 1.44 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
"""

def foo():
    seq_limit = 5
    pad_left = True

    digits = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
    padded = tf.reverse(digits, [-1]).to_tensor(0)
    tensor = tf.reverse(padded, [-1])

    paddings = tf.constant([[0, 0], [2, 0]])
    final_tensor = tf.pad(tensor, paddings)

    return final_tensor


def bar():
    seq_limit = 5
    pad_left = True
    ragged = tf.ragged.constant([[3, 1, 4, 1], [], [5, 9, 2], [6], []])
    #ragged = tf.RaggedTensor.from_row_lengths(values=values, row_lengths=diff_offsets)

    # Get vector of padding lengths using tf ops like reduce_sum.
    non_zero_entries_by_row = tf.math.reduce_sum(ragged / ragged, axis=1)
    paddings = seq_limit - non_zero_entries_by_row.numpy()

    # Make zeros ragged tensor to pad our data tensor with.
    total_entries = ragged.shape[0] * seq_limit
    non_zero_entries = tf.reduce_sum(ragged / ragged).numpy()
    zeros_count = total_entries - non_zero_entries
    zeros_values = tf.zeros(shape=(int(zeros_count)), dtype=tf.dtypes.int32)
    zeros = tf.RaggedTensor.from_row_lengths(values=zeros_values, row_lengths=paddings)

    # Concatenate zeros ragged tensor with our data tensor on either the left or the right,
    # depending on either left_pad or not.
    if pad_left:
        tensor = tf.concat([zeros, ragged], axis=1).to_tensor()
    else:
        tensor = tf.concat([ragged, zeros], axis=1).to_tensor()

    return tensor 
```

`foo()` is faster, so I implemented this approach and push a commit along the lines of this approach. Are you able to see this implementation using `tf.reverse()`?